cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,381 @@
+   *
+   * @throws cudf::logic_error if either table is empty
+   * @throws cudf::logic_error if number of returned columns is 0
+   * @throws cudf::logic_error if number of elements in `right_on` and `left_on` are not equal
+   *
+   * @param[in] left             The left table
+   * @param[in] right            The right table
+   * @param[in] left_on          The column indices from `left` to join on.
+   *                             The column from `left` indicated by `left_on[i]`
+   *                             will be compared against the column from `right`
+   *                             indicated by `right_on[i]`.
+   * @param[in] right_on         The column indices from `right` to join on.
+   *                             The column from `right` indicated by `right_on[i]`
+   *                             will be compared against the column from `left`
+   *                             indicated by `left_on[i]`.
+   * @param[in] return_columns   A vector of column indices from `left` to
+   *                             include in the returned table.
+   * @param[in] mr               Device memory resource to use for device memory allocation
+   *
+   * @returns                    Result of joining `left` and `right` tables on the columns
+   *                             specified by `left_on` and `right_on`. The resulting table
+   *                             will contain `return_columns` from `left` that match in right.
+   */
+  std::unique_ptr<cudf::experimental::table> left_anti_join(cudf::table_view const& left,
+                                                            cudf::table_view const& right,
+                                                            std::vector<cudf::size_type> const& left_on,
+                                                            std::vector<cudf::size_type> const& right_on,
+                                                            std::vector<cudf::size_type> const& return_columns,
+                                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  /**
+   * @brief Type defining join comparison operators
+   *
+   * Indicates how two elements `a` and `b` compare with one and another.
+   */
+  enum class join_comparison_operator {
+    LESS_THAN,             ///< Compare a < b
+    LESS_THAN_OR_EQUAL,    ///< Compare a <= b
+    EQUAL,                 ///< Compare a == b
+    GREATER_THAN,          ///< Compare a >  b
+    GREATER_THAN_OR_EQUAL, ///< Compare a >= b
+  };
+
+  /**
+   * @brief Type defining non-equi join operations
+   */
+  struct join_operation {
+    join_comparison_operator op;                   ///< type of comparison
+    cudf::size_type          left_column_idx;      ///< index of left column to compare
+    cudf::size_type          right_column_idx;     ///< index of right column to compare
+    bool                     only_left_in_output;  ///< if true, only output the left column
+  }
```

Comment:
Style comment :  I really like this style (justification of member names).  I wish this was more common :)

Reply:
We do specify this in our style guide for enums, but we should add it for struct members (work in progress): https://docs.google.com/document/d/10Ff4VZlSsizHdF5r9Zrw7dIwlPD0yWWSmNJ6qter7r8/edit#heading=h.u9cu9e14hobq